File: selenium_mcp/tools/keys.py
"""Key press tools."""

import logging
from pydantic import BaseModel, Field

from ..tool_base import BaseTool, ToolSchema, ToolResult
from ..context import Context

logger = logging.getLogger(__name__)

class KeyPressParams(BaseModel):
    """Parameters for key press."""
    key: str = Field(description="Key to press (e.g., 'Enter', 'Tab', 'Escape', 'F1', etc.)")
    modifiers: list[str] = Field(default=[], description="Modifier keys (ctrl, alt, shift, meta)")

class KeyPressTool(BaseTool):
    """Press keyboard keys with optional modifiers."""
    
    def _create_schema(self) -> ToolSchema:
        return ToolSchema(
            name="key_press",
            description="Press keyboard keys with optional modifiers",
            input_schema=KeyPressParams,
            tool_type="destructive"
        )
    
    async def handle(self, context: Context, params: KeyPressParams) -> ToolResult:
        """Press key with modifiers."""
        driver = context.current_tab_or_die()
        
        async def key_action():
            from selenium.webdriver.common.keys import Keys
            from selenium.webdriver.common.action_chains import ActionChains
            
            actions = ActionChains(driver)
            
            # Hold down modifiers
            for modifier in params.modifiers:
                if modifier.lower() == 'ctrl':
                    actions = actions.key_down(Keys.CONTROL)
                elif modifier.lower() == 'alt':
                    actions = actions.key_down(Keys.ALT)
                elif modifier.lower() == 'shift':
                    actions = actions.key_down(Keys.SHIFT)
                elif modifier.lower() == 'meta':
                    actions = actions.key_down(Keys.META)
            
            # Press the main key
            key_mapping = {
                'Enter': Keys.ENTER,
                'Tab': Keys.TAB,
                'Escape': Keys.ESCAPE,
                'Space': Keys.SPACE,
                'ArrowUp': Keys.ARROW_UP,
                'ArrowDown': Keys.ARROW_DOWN,
                'ArrowLeft': Keys.ARROW_LEFT,
                'ArrowRight': Keys.ARROW_RIGHT,
                'F1': Keys.F1,
                'F2': Keys.F2,
                'F3': Keys.F3,
                'F4': Keys.F4,
                'F5': Keys.F5,
                'F6': Keys.F6,
                'F7': Keys.F7,
                'F8': Keys.F8,
                'F9': Keys.F9,
                'F10': Keys.F10,
                'F11': Keys.F11,
                'F12': Keys.F12,
                'Home': Keys.HOME,
                'End': Keys.END,
                'PageUp': Keys.PAGE_UP,
                'PageDown': Keys.PAGE_DOWN,
                'Delete': Keys.DELETE,
                'Backspace': Keys.BACKSPACE,
            }
            
            selenium_key = key_mapping.get(params.key, params.key)
            actions = actions.send_keys(selenium_key)
            
            # Release modifiers
            for modifier in params.modifiers:
                if modifier.lower() == 'ctrl':
                    actions = actions.key_up(Keys.CONTROL)
                elif modifier.lower() == 'alt':
                    actions = actions.key_up(Keys.ALT)
                elif modifier.lower() == 'shift':
                    actions = actions.key_up(Keys.SHIFT)
                elif modifier.lower() == 'meta':
                    actions = actions.key_up(Keys.META)
            
            actions.perform()
            
            modifier_text = "+".join(params.modifiers) + "+" if params.modifiers else ""
            logger.info(f"⌨️ Pressed key: {modifier_text}{params.key}")
        
        # Robot Framework code
        modifier_text = "+".join(params.modifiers) + "+" if params.modifiers else ""
        code = [
            f"# Press key: {modifier_text}{params.key}",
            f"Press Keys    None    {modifier_text}{params.key}"
        ]
        
        return ToolResult(
            code=code,
            action=key_action,
            capture_snapshot=True,
            wait_for_network=False
        )

File: selenium-mcp-server/src/tools/input/keys.ts
import { z } from 'zod';
import { Key } from 'selenium-webdriver';
import { BaseTool } from '../base.js';
import { Context } from '../../context.js';
import { ToolResult } from '../../types.js';

const keyMap: Record<string, string> = {
  'enter': Key.ENTER,
  'return': Key.RETURN,
  'tab': Key.TAB,
  'escape': Key.ESCAPE,
  'esc': Key.ESCAPE,
  'backspace': Key.BACK_SPACE,
  'delete': Key.DELETE,
  'space': Key.SPACE,
  'up': Key.UP,
  'down': Key.DOWN,
  'left': Key.LEFT,
  'right': Key.RIGHT,
  'home': Key.HOME,
  'end': Key.END,
  'pageup': Key.PAGE_UP,
  'pagedown': Key.PAGE_DOWN,
  'f1': Key.F1,
  'f2': Key.F2,
  'f3': Key.F3,
  'f4': Key.F4,
  'f5': Key.F5,
  'f6': Key.F6,
  'f7': Key.F7,
  'f8': Key.F8,
  'f9': Key.F9,
  'f10': Key.F10,
  'f11': Key.F11,
  'f12': Key.F12
};

const schema = z.object({
  key: z.string().describe('Key to press (e.g., "enter", "tab", "escape", "backspace", "up", "down")'),
  ref: z.string().optional().describe('Element reference to send key to (optional, sends to active element if not specified)'),
  modifiers: z.array(z.enum(['ctrl', 'alt', 'shift', 'meta'])).optional().describe('Modifier keys to hold')
});

export class KeyPressTool extends BaseTool {
  readonly name = 'key_press';
  readonly description = 'Press a keyboard key, optionally with modifiers (ctrl, alt, shift, meta)';
  readonly inputSchema = schema;

  async execute(context: Context, params: unknown): Promise<ToolResult> {
    const { key, ref, modifiers } = this.parseParams(schema, params);

    const driver = await context.getDriver();
    const actions = driver.actions({ async: true });

    // Get the key to press
    const keyToPress = keyMap[key.toLowerCase()] || key;

    // Add modifier keys
    if (modifiers?.includes('ctrl')) {
      actions.keyDown(Key.CONTROL);
    }
    if (modifiers?.includes('alt')) {
      actions.keyDown(Key.ALT);
    }
    if (modifiers?.includes('shift')) {
      actions.keyDown(Key.SHIFT);
    }
    if (modifiers?.includes('meta')) {
      actions.keyDown(Key.META);
    }

    // If ref provided, click on element first
    if (ref) {
      const element = await context.getElementByRef(ref);
      await element.click();
    }

    // Press the key
    actions.sendKeys(keyToPress);

    // Release modifier keys
    if (modifiers?.includes('meta')) {
      actions.keyUp(Key.META);
    }
    if (modifiers?.includes('shift')) {
      actions.keyUp(Key.SHIFT);
    }
    if (modifiers?.includes('alt')) {
      actions.keyUp(Key.ALT);
    }
    if (modifiers?.includes('ctrl')) {
      actions.keyUp(Key.CONTROL);
    }

    await actions.perform();

    const modifierStr = modifiers?.length ? `${modifiers.join('+')}+` : '';
    return this.success(`Pressed ${modifierStr}${key}`, true);
  }
}
